src/lib.rs

```diff
@@ -1446,31 +1442,46 @@ impl Api {
 
 impl Drop for Api {
     fn drop(&mut self) {
-        unsafe { MsQuicClose(self.table) };
+        if self.table != ptr::null() {
+            unsafe { MsQuicClose(self.table) };
+        }
+    }
+}
+
+pub static mut API: LazyLock<Api> = LazyLock::new(|| {
```

Comment:
I assume this means it actually uses a Lock every time the API is accessed now?

Reply:
No. I suppose that we can access API in paralell from more than one threads after initiliazation. I assume LazyLock acquires "lock" only when it executes the closure for initilization. 

https://github.com/rust-lang/rust/blob/6d9f6ae36ae1299d6126ba40c15191f7aa3b79d8/library/std/src/sync/lazy_lock.rs#L207-L224

https://github.com/rust-lang/rust/blob/6d9f6ae36ae1299d6126ba40c15191f7aa3b79d8/library/std/src/sync/once.rs#L152-L158

Reply:
But usually, you need a lock to check for initialization too.

Reply:
Yes. But I suppose that the cost is cheap because an `AtomicPtr` is used to check for initilization.
https://github.com/rust-lang/rust/blob/6d9f6ae36ae1299d6126ba40c15191f7aa3b79d8/library/std/src/sys/sync/once/queue.rs#L125C1-L129C62

Reply:
But it imposes a cost on every single usage of the variable, which might be happening on a LOT of parallel threads. Imagine a machine with a 100+ CPUs, all grabbing the lock every time they want to check if the variable is initialized.

I think something a bit better is necessary. Only when you are initializing a registration _might_ you not have an initialized library after; but once you have a registration, you **know** you have initialized the library. So I'm wondering if we have (1) an unprotected global variable for the library and (2) a lazy 'run once' call when RegistrationOpen is called to make sure the global variable is properly set.

Reply:
The first time init is expensive and requires lock. After that is should be cheap. But if it is just an atomic var comparison, it   is still expensive. (Does it messup the cpu cache? with atomic acquire Relaxed?)

This is similar to cpp11 thread save static var init: https://en.cppreference.com/w/cpp/language/storage_duration#Static_block_variables
It can do none-atomic bool check.

Reply:
You still need to have a volatile read, where the only real place you need synchronization is the initialization point of all the callers (i.e. where they call RegistrationOpen). I want to make sure we don't introduce overhead here.

Reply:
We initialize APITABLE when we first initialize Api or Registration using `std::sync::Once`. There is no longer overhead when accessing APITABLE!